evm/src/executor/abi.rs

```diff
@@ -680,6 +680,266 @@ pub static HARDHAT_CONSOLE_SELECTOR_PATCHES: Lazy<HashMap<Selector, Selector>> =
     ])
 });
 
+// Similar to hardhat's but ignores all specifiers except %s. %s works on all solidity values.
+// Support for other specifiers may be added in the future
+fn format_hardhat_log(spec: &str, v: &[&str]) -> String {
+    let mut result = spec.to_string().replace("%%", "%");
+
+    let mut end = v.len();
+    for (pos, t) in v.iter().enumerate() {
+        if result.find("%s").is_some() {
+            result = result.replacen("%s", t, 1);
+        } else {
+            end = pos;
+            break
+        }
+    }
+    for t in &v[end..] {
+        result.push_str(&format!(" {}", t));
+    }
+    result
+}
+
+pub trait HardhatConsoleLogf {
+    fn logf(&self) -> String;
+}
+
+trait ToStringer {}
+impl ToStringer for String {}
+impl ToStringer for bool {}
+impl ToStringer for U256 {}
+
+trait VerboseToString {
```

Comment:
For this you might be able to just use the `UiFmt` trait here: https://github.com/foundry-rs/foundry/blob/master/common/src/fmt.rs

Reply:
yeah let's use the UiFmt trait that's also what `ToString` uses